icu4c/source/common/mlbe.cpp

```diff
@@ -53,30 +48,42 @@ int32_t MlBreakEngine::divideUpRange(UText *inText, int32_t rangeStart, int32_t
         return 0;
     }
     int32_t numBreaks = 0;
-    UnicodeString index;
-    // The ML model groups six char to evaluate if the 4th char is a breakpoint.
-    // Like a sliding window, the elementList removes the first char and appends the new char from
-    // inString in each iteration so that its size always remains at six.
-    UChar32 elementList[6];
-
-    int32_t codeUts = initElementList(inString, elementList, status);
-    int32_t length = inString.countChar32();
+    int32_t codePointLength = inString.countChar32();
+    // The ML algorithm groups six char and evaluates whether the 4th char is a breakpoint.
+    // In each iteration, it evaluates the 4th char and then moves forward one char like a sliding
+    // window. Initially, the first six values in the indexList are [-1, -1, 0, 1, 2, 3]. After
+    // moving forward, finally the last six values in the indexList are
+    // [length-4, length-3, length-2, length-1, -1, -1]. The "+4" here means four extra "-1".
+    int32_t indexSize = codePointLength + 4;
+    int32_t *indexList = (int32_t *)uprv_malloc(indexSize * sizeof(int32_t));
+    if (indexList == NULL) {
+        status = U_MEMORY_ALLOCATION_ERROR ;
+        return 0;
+    }
+    int32_t numCodeUnits = initIndexList(inString, indexList, status);
 
     // Add a break for the start.
     boundary.addElement(0, status);
     numBreaks++;
     if (U_FAILURE(status)) return 0;
 
-    for (int32_t i = 1; i < length && U_SUCCESS(status); i++) {
-        evaluateBreakpoint(elementList, i, numBreaks, boundary, status);
-        if (i + 1 >= inString.countChar32()) break;
-        // Remove the first element and append a new element
-        uprv_memmove(elementList, elementList + 1, 5 * sizeof(UChar32));
-        elementList[5] = inString.countChar32(0, codeUts) < length ? inString.char32At(codeUts) : INVALID;
-        if (elementList[5] != INVALID) {
-            codeUts += U16_LENGTH(elementList[5]);
+    for (int32_t idx = 0, breakpoint = 1; breakpoint < codePointLength && U_SUCCESS(status);
+         idx++, breakpoint++) {
+
+        numBreaks = evaluateBreakpoint(inString, indexList, idx, breakpoint, numCodeUnits,
+                                       numBreaks, boundary, status);
+        if (breakpoint + 3 < codePointLength) {
+            indexList[idx + 6] = numCodeUnits;
+            numCodeUnits += U16_LENGTH(inString.char32At(indexList[idx + 6]));
+        } else {
+            if (idx + 6 < indexSize) {
+                indexList[idx + 6] = -1;
+            }
         }
     }
+    if (indexList != NULL) {
```

Comment:
You do not need to have nullptr check here
1. if it is nullptr, you already return 0 in line 61
2. also while it was an issue in very old C compiler, C++ standard define free will be noop if the input is nullptr
see https://cplusplus.com/reference/cstdlib/free/ 
"If ptr is a null pointer, the function does nothing."
It was an issue to wrap with if NULL check years ago. but no longer needed these days

Reply:
Done. Thanks